rapids-cmake/cpm/generate_pinned_versions.cmake

```diff
@@ -0,0 +1,107 @@
+#=============================================================================
+# Copyright (c) 2024, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#=============================================================================
+include_guard(GLOBAL)
+
+#[=======================================================================[.rst:
+rapids_cpm_generate_pinned_versions
+-----------------------------------
+
+.. versionadded:: v24.04.00
+
+Generate a json file with all dependencies with pinned version values
+
+.. code-block:: cmake
+
+  rapids_cpm_generate_pinned_versions( OUTPUT <json_verions_output_path> )
+
+Generates a json file with all cpm dependencies with pinned version values.
+This allows for subsequent reproducible builds using the exact same state.
+
+The rapids-cmake default `versions.json` uses branch names or git tag names
+for dependencies. This is done so that projects can 'live at head' of dependencies.
+By using :cmake:command:`rapids_cpm_package_override` a project can specify a custom
+`versions.json` that specifies exact git SHA's so that projects have reproducible builds.
+
+:cmake:command:`rapids_cpm_generate_pinned_versions` can be used to transform a set of
+rapids-cmake dependencies from branch names to pinned values. This can be used in subsequent
+builds, e.g:
+
+  1. Have CI run with `versions.json` which tracks dependency by branch name
+  2. Store the generated pinned `versions.json` from the CI builds
+  3. If build is good, create the release branch and commit the generated pinned `versions.json`
+     to have reproducible builds for that release
+
+``OUTPUT``
+Specify a file path to write the pinned versions information to. Can be called multiple
+times and each unique path will be written too.
+
+The generated json file will have the following entries for each package:
+
+.. code-block:: json
+
+  {
+    "version": "<CPM_PACKAGE_<package_name>_VERSION>",
+    "git_url": "<deduced>",
+    "git_tag": "<deduced>",
+    "git_shallow": "false",
+    "always_download": "true"
```

Comment:
I mentioned above that booleans should not be quoted -- in this example, the values will be interpreted as strings with contents `"false"` or `"true"`.

Reply:
Good point, working on an update to `rapids_cpm_pinning_create_and_set_member` to detect json types correctly so we know when to quote.

Reply:
Okay updatred `rapids_cpm_pinning_create_and_set_member` and it now works around the issues reported upstream at https://gitlab.kitware.com/cmake/cmake/-/issues/25716